jobs/update_last_read/dataset.py

```diff
@@ -0,0 +1,55 @@
+import re
+from dataclasses import dataclass
+from urllib.parse import unquote
+
+STAGES = ["raw", "intermediate", "processed"]
+CONFIDENTIALITIES = ["green", "yellow", "red"]
+
+
+@dataclass
+class DatasetEntry:
+    """An S3 object entry that looks like it belongs to a dataset."""
+
+    stage: str
+    confidentiality: str
+    dataset_id: str
+    version: str
+    edition_path: str
+    filename: str
+
+    @classmethod
+    def from_s3_key(cls, key):
+        """Return a `DatasetEntry` object corresponding to `key`.
+
+        If `key` doesn't look like it belongs to a dataset, return `None`.
+        """
+        if isinstance(key, str):
+            # Amazon URL-encodes the key twice for unknown reasons, so decode
+            # it twice.
+            key_unquoted = unquote(unquote(key))
+
+            # fmt: off
+            pattern = re.compile("/".join([
+                r"(?P<stage>[^/]+)",            # Stage
+                r"(?P<confidentiality>[^/]+)",  # Confidentiality
+                r"(?P<dataset>\S+)",            # Dataset
+                r"version=(?P<version>[^/]+)",  # Version
+                r"(?P<edition_path>\S+)",       # Edition path
```

Comment:
Det har vel ikke mye å si for øyeblikket, men er det meningen at `edition_path` skal matche "resten" av stien (utover selve filnavnet), f.eks. `edition=20240728T020032/_delta_log`? Eller bør dette heller fanges opp i `filename`, f.eks. `_delta_log/00000000000000000000.json`

Matche `edition` på tilsvarende måte som `version` (inkl. `latest` så klart), og heller lagre resterende sti i `filepath`?

Reply:
Jepp, enig.